src/store/helpers.js

```diff
@@ -26,3 +28,140 @@ export function onlyHandleAPIError (error, handleFn) {
+
+export function isRequestError (error) {
+  const { response: { status = -1 } = {} } = error
+  if (status >= 500) {
+    return true
+  }
+  if (error.request) {
+    return true
+  }
+  return false
+}
+
+/**
+ * Defines a vues module that can be used to handle async request metadata.
+ *
+ * Instantiation:
+
+ const meta = defineRequestModule()
+ export const modules = { meta }
+
+ * Usage:
+
+async save ({ commit, dispatch }, group) {
+  dispatch('meta/request', {
+    id: `save-${group.id}`,
+    async run () {
+      const updatedGroup = await groups.save(group)
+      commit(types.RECEIVE_GROUP, { group: updatedGroup }) // other commits when successful
+    },
+  })
+},
+ */
+export function defineRequestModule () {
+  const types = {
+    REQUEST: 'Request',
+    RECEIVE_SUCCESS: 'Receive Success',
+    RECEIVE_ERROR: 'Receive Error',
+    CLEAR: 'Clear',
+    CLEAR_ALL: 'Clear all',
+  }
+
+  function initialState () {
+    return {
+      entries: {},
+    }
+  }
+
+  const state = initialState()
+
+  const getters = {
+    get: state => id => state.entries[id] || ({
+      isWaiting: false,
+      error: undefined,
+      success: false,
+    }),
+    error: (state, getters) => id => field => {
+      const { error: { [field]: [ message ] = [] } = {} } = getters.get(id)
+      return message
+    },
+    isWaiting: (state, getters) => id => getters.get(id).isWaiting,
+    success: (state, getters) => id => getters.get(id).success,
+  }
+
+  const actions = {
+    async request ({ commit }, { id, run, onValidationError }) {
+      commit(types.REQUEST, { id })
+      try {
+        await run()
+        commit(types.RECEIVE_SUCCESS, { id })
+      }
+      catch (error) {
+        if (isValidationError(error)) {
+          commit(types.RECEIVE_ERROR, { id, error: error.response.data })
+          onValidationError(error)
+        }
+        else if (isRequestError(error)) {
+          // TODO: should we commit this at all?
+          commit(types.RECEIVE_ERROR, { id, error: error.request })
+        }
+        else {
+          // some other error, can't handle it here
+          throw error
+        }
+      }
+    },
+    clear ({ commit }) {
+      commit('clear')
+    },
+  }
+
+  const mutations = {
+    [types.REQUEST] (state, { id }) {
+      Vue.set(state.entries, id, {
+        isWaiting: true,
+        error: undefined,
+        success: false,
+      })
+    },
+    [types.RECEIVE_SUCCESS] (state, { id }) {
+      Vue.set(state.entries, id, {
+        isWaiting: false,
+        error: undefined,
+        success: true,
+      })
+    },
+    [types.RECEIVE_ERROR] (state, { id, error }) {
+      Vue.set(state.entries, id, {
+        isWaiting: false,
+        error,
+        success: false,
+      })
+    },
+    [types.CLEAR] (state, { id }) {
+      Vue.delete(state.entries, id)
+    },
+    [types.CLEAR_ALL] (state) {
+      Object.entries(initialState())
+        .forEach(([prop, value]) => Vue.set(state, prop, value))
```

Comment:
I think `Object.assign(state, initialState())` should work fine here, as the props in the initial state, are by definition already defined... and object.assign will just assign them again.

See https://vuejs.org/v2/guide/reactivity.html#Change-Detection-Caveats (implicitly says it works fine)